src/main/generic/consensus/Policy.js

```diff
@@ -36,11 +36,140 @@ class Policy {
+     * @return {number}
+     */
+    static _supplyAfter(initialSupply, blockHeight, startHeight=0) {
+        let supply = initialSupply;
+        for (let i = startHeight; i <= blockHeight; ++i) {
+            supply += Policy._blockRewardAt(supply, i);
+        }
+        return supply;
+    }
+
+    /**
+     * Circulating supply after block.
+     * @param {number} blockHeight
+     * @return {number}
+     */
+    static supplyAfter(blockHeight) {
+        // FIXME: Change for main net.
+        if (blockHeight < Policy.EMISSION_CURVE_START) {
+            return Policy.INITIAL_SUPPLY + (blockHeight+1) * Policy.coinsToSatoshis(5);
+        }
+        // Luna net supply after start of emission curve.
+        const initialSupply = Policy.INITIAL_SUPPLY + Policy.EMISSION_CURVE_START * Policy.coinsToSatoshis(5);
+
+        // Calculate last entry in supply cache that is below blockHeight.
+        let startHeight = Math.floor(blockHeight / Policy._supplyCacheInterval) * Policy._supplyCacheInterval;
+        startHeight = Math.max(/* FIXME change to 0 for main net */ Policy.EMISSION_CURVE_START, Math.min(startHeight, Policy._supplyCacheMax));
+
+        // Calculate respective block for the last entry of the cache and the targeted height.
+        const startI = startHeight / Policy._supplyCacheInterval;
+        const endI = Math.floor(blockHeight / Policy._supplyCacheInterval);
+
+        // The starting supply is the initial supply at the beginning and a cached value afterwards.
+        let supply = startHeight === /* FIXME change to 0 for main net */ Policy.EMISSION_CURVE_START ? initialSupply : Policy._supplyCache.get(startHeight);
+        // Use and update cache.
+        for (let i=startI; i<endI; ++i) {
+            startHeight = i * Policy._supplyCacheInterval;
+            // Since the cache stores the supply *before* a certain block, subtract one.
+            const endHeight = (i+1) * Policy._supplyCacheInterval - 1;
+            supply = Policy._supplyAfter(supply, endHeight, startHeight);
+            // Don't forget to add one again.
+            Policy._supplyCache.set(endHeight + 1, supply);
+            Policy._supplyCacheMax = endHeight + 1;
+        }
+
+        // Calculate remaining supply (this also adds the block reward for endI*interval).
+        return Policy._supplyAfter(supply, blockHeight, endI*Policy._supplyCacheInterval);
+    }
+
+    /**
+     * Miner reward per block.
+     * @param {number} currentSupply
+     * @param {number} blockHeight
+     * @return {number}
+     */
+    static _blockRewardAt(currentSupply, blockHeight) {
+        const remaining = Policy.TOTAL_SUPPLY - currentSupply;
+        if (blockHeight >= Policy.EMISSION_TAIL_START && remaining >= Policy.EMISSION_TAIL_REWARD) {
+            return Policy.EMISSION_TAIL_REWARD;
+        }
+        const remainder = remaining % Policy.EMISSION_SPEED;
+        return (remaining-remainder) / Policy.EMISSION_SPEED;
```

Comment:
Aren't these two lines the same as
```
return Math.floor(remaining / Policy.EMISSION_SPEED);
```

Reply:
In principle, yes, except that the `Math.floor` approach uses a float division. If remaining is large and the division yields a lot of decimal places, floating point inaccuracies might occur.
Our code ensures that the value is divisible by the emission speed and prevents inaccuracies.